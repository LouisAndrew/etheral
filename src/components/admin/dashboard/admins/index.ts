
import { Admins } from './admins';

export default Admins;