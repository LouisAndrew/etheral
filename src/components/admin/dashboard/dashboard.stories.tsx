import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Dashboard from '.';
// import more addons

const story = storiesOf('Components.Admin.Dashboard', module);

story.add('Component', () => <Dashboard logout={action('logout')} />);
