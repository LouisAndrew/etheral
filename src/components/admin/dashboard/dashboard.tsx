import React, { useState } from 'react';
import { Flex, Box } from 'rebass';

import Orders from './orders';
import Navigation from './navigation';

type Props = {
    db: firebase.firestore.Firestore;
    logout: () => void;
};

export enum StateViews {
    ORDERS,
    ADMINS,
    LINKS,
    NONE,
}

const Dashboard: React.FC<Props> = ({ logout, db }) => {
    const [view, setView] = useState<StateViews>(StateViews.NONE);

    const changeView = (viewName: StateViews) => {
        setView(viewName);
    };

    // eslint-disable-next-line immutable/no-let, @typescript-eslint/tslint/config
    let toRender;
    switch (view) {
        case StateViews.ORDERS:
            toRender = <Orders db={db} />;
            break;
        case StateViews.ADMINS:
            toRender = <h1>Admins</h1>;
            break;
        case StateViews.LINKS:
            toRender = <h1>Links</h1>;
            break;
        case StateViews.NONE:
            toRender = <h1>None</h1>;
            break;
        default:
            toRender = <h1>None</h1>;
            break;
    }

    return (
        <Flex data-testid="dashboard" min-height="100vh">
            {/* Navigation */}
            <Box sx={{ position: 'sticky' }}>
                <Navigation
                    logout={logout}
                    changeView={changeView}
                    inView={view}
                />
            </Box>
            <Box width="100%">{toRender}</Box>
        </Flex>
    );
};

export { Dashboard };
