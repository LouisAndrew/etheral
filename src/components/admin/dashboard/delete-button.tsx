import React from 'react';

import { Box, Text } from 'rebass';
import { InlineIcon } from '@iconify/react';
import deleteBinFill from '@iconify/icons-ri/delete-bin-fill';

type Props = {
    onClick: any;
};

const DeteleButton: React.FC<Props> = ({ onClick }) => {
    const textStyling = {
        fontFamily: 'body',
        fontSize: [1],
        // color: 'black.0' => removed to be able to apply custom color on remove text,
        fontWeight: 'medium',
        css: `
			 white-space: nowrap;
			overflow: hidden;
			text-overflow: ellipsis;
		`,
    };

    return (
        <Box width="100%">
            <Text
                {...textStyling}
                p={[1]}
                width="fit-content"
                sx={{
                    display: 'flex',
                    alignItems: 'center',
                    color: 'misc.discount',
                    bg: 'transparent',
                    transition: '0.2s',
                    svg: {
                        mr: 2,
                        path: {
                            fill: 'misc.discount',
                            transition: '0.2s',
                        },
                    },
                    '&:hover': {
                        cursor: 'pointer',
                        bg: 'misc.discount',
                        color: '#fff',
                        'svg path': { fill: '#fff' },
                    },
                }}
                onClick={onClick}
                role="button"
            >
                <InlineIcon icon={deleteBinFill} />
                Remove
            </Text>
        </Box>
    );
};

export default DeteleButton;
