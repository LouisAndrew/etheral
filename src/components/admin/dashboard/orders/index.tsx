import React, { useEffect, useState } from 'react';
import firebase from 'gatsby-plugin-firebase';
import { set } from 'lodash';

import { Orders as OrdersEl } from './orders';
import { Order } from 'helper/schema/order';

type Props = {
    db: firebase.firestore.Firestore;
};

const Orders: React.FC<Props> = ({ db }) => {
    const [orders, setOrders] = useState<Order[] | undefined>(undefined);

    useEffect(() => {
        (async () => {
            // eslint-disable-next-line @typescript-eslint/tslint/config, immutable/no-let
            let ordersContainer: Order[] = [];
            const reqOrders = await db.collection('order').get();

            await reqOrders.forEach(doc => {
                const orderData: Order = doc.data() as Order;
                set(
                    orderData,
                    'date',
                    new Date((orderData.date as any).seconds * 1000)
                );

                if (orderData.shippingData) {
                    set(
                        orderData,
                        'shippingData.shippingDate',
                        new Date(
                            (orderData.shippingData.shippingDate as any)
                                .seconds * 1000
                        )
                    );
                }

                ordersContainer = [...ordersContainer, orderData];
            });

            setOrders(await ordersContainer);
        })();
    }, []);

    return orders ? <OrdersEl orders={orders} db={db} /> : <></>;
};

export default Orders;
