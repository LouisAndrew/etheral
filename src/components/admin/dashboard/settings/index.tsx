/* eslint-disable prefer-const */
/* eslint-disable @typescript-eslint/tslint/config, immutable/no-let */

import React, { useState, useEffect } from 'react';
import firebase from 'gatsby-plugin-firebase';

import { Settings as SettingsEl } from './settings';

export enum Status {
    NEW,
    UPDATE,
}

export type Timestamps = {
    lastModifiedDate?: Date;
    createdDate: Date;
    status: Status;
};

export type LatestProducts = Timestamps & {
    name: string;
    pid: string;
};

export type LatestCollections = Timestamps & {
    name: string;
    cid: string;
};

export type LatestBlogs = Timestamps & {
    slug: string;
};

const Settings: React.FC<{}> = () => {
    const [db, setDb] = useState<firebase.firestore.Firestore | null>(null);

    const [isLoading, setIsLoading] = useState(false);

    const [latestProducts, setLatestProducts] = useState<LatestProducts[]>([]);
    const [latestCollections, setLatestCollections] = useState<
        LatestCollections[]
    >([]);
    const [latestBlogs, setLatestBlogs] = useState<LatestBlogs[]>([]);
    // const [latestHomepage, setLatestHomepage] = useState<
    //     LatestHomepage | undefined
    // >(undefined);

    useEffect(() => {
        setDb(firebase.firestore());
    }, []);

    // DO NOT OPEN THIS
    // basically just a way to fetch ALL data from firestore (same as on gatsby-node.js)
    useEffect(() => {
        if (db) {
            fetchData();
        }
    }, [db]);

    const fetchData = async () => {
        if (db) {
            await setIsLoading(true);

            const contentRef = db.collection('fl_content');

            let lastUpdate = 0;

            const lastUpdateRef = db
                .collection('website-update')
                .orderBy('lastUpdate', 'desc')
                .limit(1);

            await lastUpdateRef.get().then(docs =>
                docs.forEach(async doc => {
                    lastUpdate = await doc.data().lastUpdate.seconds;
                })
            );

            // setting all refs here.
            const productsRef = contentRef.where(
                '_fl_meta_.schema',
                '==',
                'product'
            );
            const collectionsRef = contentRef.where(
                '_fl_meta_.schema',
                '==',
                'collection'
            );

            // const homepageRef = contentRef.where(
            //     '_fl_meta_.schema',
            //     '==',
            //     'homepage'
            // );
            const blogsRef = contentRef.where('_fl_meta_.schema', '==', 'blog');

            let collectionDocs: firebase.firestore.DocumentData[] = [];
            let productDocs: firebase.firestore.DocumentData[] = [];
            let blogDocs: firebase.firestore.DocumentData[] = [];

            // not sure why, but whenever i fetched homepage data, latest products and collections won't be fetched

            // await homepageRef // fetch homepage data
            //     .get()
            //     .then(docs => {
            //         if (docs.size > 0) {
            //             docs.forEach(doc => {
            //                 // homepageDoc = doc.data();
            //                 if (doc) {
            //                     const data = doc.data();

            //                     if (
            //                         data._fl_meta_.createdDate.seconds >
            //                             lastUpdate ||
            //                         data._fl_meta_.lastModifiedDate.seconds >
            //                             lastUpdate
            //                     ) {
            //                         setLatestHomepage({
            //                             createdDate: data._fl_meta_.createdDate.toDate(),
            //                             lastModifiedDate: data._fl_meta_.lastModifiedDate.toDate(),
            //                             status: Status.UPDATE,
            //                         });
            //                     }
            //                 }
            //             });
            //         }
            //     });

            // fetch all products
            await productsRef.get().then(docs =>
                docs.forEach(doc => {
                    productDocs = [...productDocs, doc.data()];
                })
            );

            // fetch all collections
            await collectionsRef.get().then(docs =>
                docs.forEach(doc => {
                    collectionDocs = [...collectionDocs, doc.data()];
                })
            );

            // fetch all blogs
            await blogsRef.get().then(docs =>
                docs.forEach(doc => {
                    blogDocs = [...blogDocs, doc.data()];
                })
            );

            await setLatestProducts(
                productDocs
                    // filter function to filter latest product that has been created OR updated after the last update count.
                    .filter(
                        product =>
                            product._fl_meta_.createdDate.seconds >
                                lastUpdate ||
                            (product._fl_meta_.lastModifiedDate &&
                                product._fl_meta_.lastModifiedDate.seconds >
                                    lastUpdate)
                    )
                    .map(product => ({
                        pid: product.id,
                        name: product.name,
                        lastModifiedDate: product._fl_meta_.lastModifiedDate
                            ? product._fl_meta_.lastModifiedDate.toDate()
                            : undefined,
                        createdDate: product._fl_meta_.createdDate.toDate(),
                        status:
                            product._fl_meta_.createdDate.seconds > lastUpdate // no need to check if created date is smaller than last update
                                ? Status.NEW
                                : Status.UPDATE,
                    }))
            );

            await setLatestCollections(
                collectionDocs
                    .filter(
                        collection =>
                            collection._fl_meta_.createdDate.seconds >
                                lastUpdate ||
                            (collection._fl_meta_.lastModifiedDate &&
                                collection._fl_meta_.lastModifiedDate.seconds >
                                    lastUpdate)
                    )
                    .map(collection => ({
                        cid: collection.id,
                        name: collection.name,
                        lastModifiedDate: collection._fl_meta_.lastModifiedDate
                            ? collection._fl_meta_.lastModifiedDate.toDate()
                            : undefined,
                        createdDate: collection._fl_meta_.createdDate.toDate(),
                        status:
                            collection._fl_meta_.createdDate.seconds >
                            lastUpdate
                                ? Status.NEW
                                : Status.UPDATE,
                    }))
            );

            await setLatestBlogs(
                blogDocs
                    .filter(
                        blog =>
                            blog._fl_meta_.createdDate.seconds > lastUpdate ||
                            (blog._fl_meta_.lastModifiedDate &&
                                blog._fl_meta_.lastModifiedDate.seconds >
                                    lastUpdate)
                    )
                    .map(blog => ({
                        slug: blog.slug,
                        lastModifiedDate: blog._fl_meta_.lastModifiedDate
                            ? blog._fl_meta_.lastModifiedDate.toDate()
                            : undefined,
                        createdDate: blog._fl_meta_.createdDate.toDate(),
                        status:
                            blog._fl_meta_.createdDate.seconds > lastUpdate
                                ? Status.NEW
                                : Status.UPDATE,
                    }))
            );

            await setIsLoading(false);
        }
    };

    if (db && !isLoading) {
        return (
            <SettingsEl
                latestProducts={latestProducts}
                latestCollections={latestCollections}
                latestBlogs={latestBlogs}
            />
        );
    }

    return <h1>Wait, loading</h1>;
};

export default Settings;
