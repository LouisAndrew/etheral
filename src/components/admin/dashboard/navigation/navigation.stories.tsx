import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
// import more addons

import Navigation from '.';
import { StateViews } from '../dashboard';

const story = storiesOf('Components.Admin.Dashboard.Navigation', module);

story
    .add('Component with no active view', () => (
        <Navigation
            inView={StateViews.NONE}
            logout={action('log-out')}
            changeView={viewName => {
                action(`changing-view: ${viewName}`);
            }}
        />
    ))
    .add('Component with active view on orders', () => (
        <Navigation
            inView={StateViews.ORDERS}
            logout={action('log-out')}
            changeView={viewName => {
                action(`changing-view: ${viewName}`);
            }}
        />
    ));
