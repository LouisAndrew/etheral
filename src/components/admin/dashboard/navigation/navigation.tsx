import React from 'react';
import { Flex, Box, Button } from 'rebass';
import { InlineIcon } from '@iconify/react';
import linksLine from '@iconify/icons-ri/links-line';
import fileList3Line from '@iconify/icons-ri/file-list-3-line';
import user3Line from '@iconify/icons-ri/user-3-line';
import logoutBoxLine from '@iconify/icons-ri/logout-box-line';

import { StateViews } from '../dashboard';

type Props = {
    inView: StateViews;
    logout: () => void;
    changeView: (viewname: StateViews) => void;
};

const Navigation: React.FC<Props> = ({ inView, logout, changeView }) => {
    return (
        <Flex flexDirection="column" alignItems="flex-start" height="100vh">
            <Button
                variant={
                    inView === StateViews.ORDERS
                        ? 'adminLinkActive'
                        : 'adminLink'
                }
                onClick={() => {
                    changeView(StateViews.ORDERS);
                }}
            >
                <InlineIcon icon={fileList3Line} />
                Orders
            </Button>
            <Button
                variant={
                    inView === StateViews.LINKS
                        ? 'adminLinkActive'
                        : 'adminLink'
                }
                onClick={() => {
                    changeView(StateViews.LINKS);
                }}
            >
                <InlineIcon icon={linksLine} />
                Links
            </Button>
            <Button
                variant={
                    inView === StateViews.ADMINS
                        ? 'adminLinkActive'
                        : 'adminLink'
                }
                onClick={() => {
                    changeView(StateViews.ADMINS);
                }}
            >
                <InlineIcon icon={user3Line} />
                Admin users
            </Button>
            <Box my="auto" />
            <Button variant={'adminLink'} onClick={logout}>
                <InlineIcon icon={logoutBoxLine} />
                Log out
            </Button>
        </Flex>
    );
};

export { Navigation };
