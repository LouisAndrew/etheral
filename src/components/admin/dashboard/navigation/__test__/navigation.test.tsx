import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';

import { Navigation } from '../navigation';
import { StateViews } from '../../dashboard';

describe('Admin navigation component', () => {
    let viewContainer: StateViews = StateViews.NONE;

    const mockLogout = jest.fn(() => {});
    const mockChangeView = jest.fn((viewName: StateViews) => {
        viewContainer = viewName;
    });

    const el: React.ReactElement = (
        <Navigation
            inView={viewContainer}
            logout={mockLogout}
            changeView={mockChangeView}
        />
    );

    afterEach(cleanup);
    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(el, div);
    });

    it('should invoke logout function when logout button is clicked', () => {
        const { getByText } = render(el);

        const LogoutButton: HTMLElement = getByText('Log out');

        userEvent.click(LogoutButton);
        expect(mockLogout).toBeCalled();
    });

    it('should call changeView function accordingly ', () => {
        const { getByText } = render(el);

        const OrderButton: HTMLElement = getByText('Orders');

        userEvent.click(OrderButton);
        expect(mockChangeView).toBeCalled();
        expect(viewContainer).toBe(StateViews.ORDERS);
    });

    it('matches snapshot', () => {
        const tree = renderer.create(el).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
