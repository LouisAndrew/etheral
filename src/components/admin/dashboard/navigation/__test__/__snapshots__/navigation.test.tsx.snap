// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin navigation component matches snapshot 1`] = `
<div
  className="css-1b6hbl9"
>
  <button
    className="css-vhv7r1"
    onClick={[Function]}
  >
    <svg
      dangerouslySetInnerHTML={
        Object {
          "__html": "<path d=\\"M19 22H5a3 3 0 0 1-3-3V3a1 1 0 0 1 1-1h14a1 1 0 0 1 1 1v12h4v4a3 3 0 0 1-3 3zm-1-5v2a1 1 0 0 0 2 0v-2h-2zm-2 3V4H4v15a1 1 0 0 0 1 1h11zM6 7h8v2H6V7zm0 4h8v2H6v-2zm0 4h5v2H6v-2z\\" fill=\\"currentColor\\"/>",
        }
      }
      focusable={false}
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "transform": "rotate(360deg)",
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    />
    Orders
  </button>
  <button
    className="css-vhv7r1"
    onClick={[Function]}
  >
    <svg
      dangerouslySetInnerHTML={
        Object {
          "__html": "<path d=\\"M13.06 8.11l1.415 1.415a7 7 0 0 1 0 9.9l-.354.353a7 7 0 0 1-9.9-9.9l1.415 1.415a5 5 0 1 0 7.071 7.071l.354-.354a5 5 0 0 0 0-7.07l-1.415-1.415l1.415-1.414zm6.718 6.011l-1.414-1.414a5 5 0 1 0-7.071-7.071l-.354.354a5 5 0 0 0 0 7.07l1.415 1.415l-1.415 1.414l-1.414-1.414a7 7 0 0 1 0-9.9l.354-.353a7 7 0 0 1 9.9 9.9z\\" fill=\\"currentColor\\"/>",
        }
      }
      focusable={false}
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "transform": "rotate(360deg)",
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    />
    Links
  </button>
  <button
    className="css-vhv7r1"
    onClick={[Function]}
  >
    <svg
      dangerouslySetInnerHTML={
        Object {
          "__html": "<path d=\\"M20 22h-2v-2a3 3 0 0 0-3-3H9a3 3 0 0 0-3 3v2H4v-2a5 5 0 0 1 5-5h6a5 5 0 0 1 5 5v2zm-8-9a6 6 0 1 1 0-12a6 6 0 0 1 0 12zm0-2a4 4 0 1 0 0-8a4 4 0 0 0 0 8z\\" fill=\\"currentColor\\"/>",
        }
      }
      focusable={false}
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "transform": "rotate(360deg)",
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    />
    Admin users
  </button>
  <div
    className="css-gi5z2v"
  />
  <button
    className="css-vhv7r1"
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [
            Class {
              "_dispatchInstances": null,
              "_dispatchListeners": null,
              "_targetInst": null,
              "altKey": null,
              "bubbles": null,
              "button": null,
              "buttons": null,
              "cancelable": null,
              "clientX": null,
              "clientY": null,
              "ctrlKey": null,
              "currentTarget": [Function],
              "defaultPrevented": null,
              "detail": null,
              "dispatchConfig": null,
              "eventPhase": null,
              "getModifierState": [Function],
              "isDefaultPrevented": [Function],
              "isPropagationStopped": [Function],
              "isTrusted": null,
              "metaKey": null,
              "movementX": [Function],
              "movementY": [Function],
              "nativeEvent": null,
              "pageX": null,
              "pageY": null,
              "relatedTarget": [Function],
              "screenX": null,
              "screenY": null,
              "shiftKey": null,
              "target": null,
              "timeStamp": [Function],
              "type": null,
              "view": null,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <svg
      dangerouslySetInnerHTML={
        Object {
          "__html": "<path d=\\"M4 18h2v2h12V4H6v2H4V3a1 1 0 0 1 1-1h14a1 1 0 0 1 1 1v18a1 1 0 0 1-1 1H5a1 1 0 0 1-1-1v-3zm2-7h7v2H6v3l-5-4l5-4v3z\\" fill=\\"currentColor\\"/>",
        }
      }
      focusable={false}
      height="1em"
      preserveAspectRatio="xMidYMid meet"
      style={
        Object {
          "transform": "rotate(360deg)",
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    />
    Log out
  </button>
</div>
`;
